Phoenix/PHSpace.m

```diff
@@ -61,8 +65,25 @@ @implementation PHSpace
 void CGSRemoveWindowsFromSpaces(CGSConnectionID connection, CFArrayRef windowIds, CFArrayRef spaceIds);
 
 // XXX: Undocumented private API to move the given windows (CGWindowIDs) to the given space
+//      Only works prior to MacOS 14.5!
 void CGSMoveWindowsToManagedSpace(CGSConnectionID connection, CFArrayRef windowIds, CGSSpaceID spaceId);
 
+/**
+ * The two functions below serve as a workaround to CGSMoveWindowsToManagedSpace breaking in MacOS 14.5.
+ * - https://github.com/kasper/phoenix/issues/348
+ * - https://github.com/koekeishiya/yabai/issues/2240#issuecomment-2116326165
+ * - https://github.com/ianyh/Amethyst/issues/1643#issuecomment-2132519682
+ */
+
+// XXX: Undocumented private API to set a space's (legacy) compatID
+CGError CGSSpaceSetCompatID(CGSConnectionID connection, CGSSpaceID spaceId, NSUInteger compatID);
+
+// XXX: Undocumented private API to move the given windows (CGWindowIDs) to the given space by (legacy) compatID
+CGError CGSSetWindowListWorkspace(CGSConnectionID connection,
+                                  CGWindowID *windowIds,
```

Comment:
Is this a single window ID or an array ref?

Reply:
Or well I guess it’s the ID sequence, but the type or naming seems confusing.

Reply:
My understanding is that it's a pointer to an int ID array, which in C would just be a contiguous block of memory. I'm not sure what the idiomatic way of naming and converting/passing it into the API is in Objective C, so I'm happy to make any change you think will improve clarity!

Reply:
Changed to `CGWindowID windowIds[]` since I think that is at least clearer.

Reply:
Out of curiosity, could you try making this `CFArrayRef windowIds`, then you could also try just to create a normal mutable array (https://github.com/kasper/phoenix/blob/b8f8f277a8a4435aff265fdc58135233c2310eef/Phoenix/PHSpace.m#L189) and bridge it as an array ref (https://github.com/kasper/phoenix/blob/b8f8f277a8a4435aff265fdc58135233c2310eef/Phoenix/PHSpace.m#L209). Would this work?

Reply:
I tried this:

```objectivec
- (void)moveWindowsWithCompatId:(NSArray<PHWindow *> *)windows {
    NSArray<NSNumber *> *ids = [self identifiersForWindows:windows];
    CGSConnectionID connection = CGSMainConnectionID();
    CGSSpaceSetCompatID(connection, self.identifier, MoveWindowsCompatId);
    CGSSetWindowListWorkspace(connection, (__bridge CFArrayRef)ids, [ids count], MoveWindowsCompatId);
    CGSSpaceSetCompatID(connection, self.identifier, 0x0);
}
```

Unfortunately, it doesn't work - I think the API expects a C-style array where values are stored in contiguous memory, and `NS(Mutable)Array` have no such guarantees.

Reply:
Yep, I was just mostly curious. Thanks! 😄